File: recipes/subms-lsm-tree/java/src/main/java/com/submillisecond/recipes/lsm/Memtable.java
package com.submillisecond.recipes.lsm;

import java.util.Optional;
import java.util.TreeMap;

/**
 * In-memory buffer of pending writes, sorted by key. A {@code null} value
 * represents a tombstone - present in the map, marking the key as deleted.
 */
final class Memtable {

    private final TreeMap<String, byte[]> entries = new TreeMap<>();
    private int approxSizeBytes = 0;

    void put(String key, byte[] value) {
        byte[] prev = entries.put(key, value);
        if (prev == null) {
            approxSizeBytes += keyCost(key) + valueCost(value);
        } else {
            approxSizeBytes += valueCost(value) - valueCost(prev);
        }
    }

    void delete(String key) {
        put(key, null);
    }

    /**
     * @return {@link Optional#empty()} if the key is not in the memtable at all;
     *         an empty {@link Lookup} if the key is tombstoned;
     *         a {@link Lookup} carrying the value otherwise.
     */
    Optional<Lookup> get(String key) {
        if (!entries.containsKey(key)) return Optional.empty();
        return Optional.of(new Lookup(entries.get(key)));
    }

    int approxSizeBytes() {
        return approxSizeBytes;
    }

    int entryCount() {
        return entries.size();
    }

    boolean isEmpty() {
        return entries.isEmpty();
    }

    Iterable<java.util.Map.Entry<String, byte[]>> sortedEntries() {
        return entries.entrySet();
    }

    /**
     * Entries whose key is in {@code [lo, hi)} (a {@code null} bound is
     * unbounded), in key order. A tombstone surfaces as an entry with a
     * {@code null} value, resolved by the caller.
     */
    Iterable<java.util.Map.Entry<String, byte[]>> range(String lo, String hi) {
        java.util.NavigableMap<String, byte[]> view;
        if (lo == null && hi == null) {
            view = entries;
        } else if (lo == null) {
            view = entries.headMap(hi, false);
        } else if (hi == null) {
            view = entries.tailMap(lo, true);
        } else {
            view = entries.subMap(lo, true, hi, false);
        }
        return view.entrySet();
    }

    void clear() {
        entries.clear();
        approxSizeBytes = 0;
    }

    private static int keyCost(String key) {
        return key == null ? 0 : key.length();
    }

    private static int valueCost(byte[] value) {
        return value == null ? 1 : value.length;
    }

    record Lookup(byte[] value) {
        boolean isTombstone() {
            return value == null;
        }
    }
}

File: recipes/subms-lsm-tree/rust/src/memtable.rs
use std::collections::BTreeMap;

/// In-memory buffer of pending writes, sorted by key. `None` represents a
/// tombstone - present in the map, marking the key as deleted.
pub(crate) struct Memtable {
    entries: BTreeMap<String, Option<Vec<u8>>>,
    approx_size_bytes: usize,
}

impl Memtable {
    pub(crate) fn new() -> Self {
        Self {
            entries: BTreeMap::new(),
            approx_size_bytes: 0,
        }
    }

    pub(crate) fn put(&mut self, key: &str, value: Option<Vec<u8>>) {
        let new_val_cost = value.as_ref().map(|v| v.len()).unwrap_or(1);
        match self.entries.insert(key.to_string(), value) {
            None => self.approx_size_bytes += key.len() + new_val_cost,
            Some(prev) => {
                let prev_cost = prev.as_ref().map(|v| v.len()).unwrap_or(1);
                self.approx_size_bytes = self.approx_size_bytes + new_val_cost - prev_cost;
            }
        }
    }

    /// Returns `None` if the key is absent, `Some(None)` if tombstoned,
    /// `Some(Some(v))` if a value is present.
    pub(crate) fn get(&self, key: &str) -> Option<Option<&[u8]>> {
        self.entries.get(key).map(|opt| opt.as_deref())
    }

    pub(crate) fn approx_size_bytes(&self) -> usize {
        self.approx_size_bytes
    }

    pub(crate) fn entry_count(&self) -> usize {
        self.entries.len()
    }

    pub(crate) fn is_empty(&self) -> bool {
        self.entries.is_empty()
    }

    pub(crate) fn sorted_entries(&self) -> impl Iterator<Item = (&str, Option<&[u8]>)> {
        self.entries.iter().map(|(k, v)| (k.as_str(), v.as_deref()))
    }

    /// Entries whose key is in `[lo, hi)` (either bound `None` = unbounded), in
    /// key order. Tombstones surface as `(key, None)`, resolved by the caller.
    pub(crate) fn range<'a>(
        &'a self,
        lo: Option<&str>,
        hi: Option<&str>,
    ) -> impl Iterator<Item = (&'a str, Option<&'a [u8]>)> {
        use std::ops::Bound;
        let low = lo.map(Bound::Included).unwrap_or(Bound::Unbounded);
        let high = hi.map(Bound::Excluded).unwrap_or(Bound::Unbounded);
        self.entries
            .range::<str, _>((low, high))
            .map(|(k, v)| (k.as_str(), v.as_deref()))
    }

    pub(crate) fn clear(&mut self) {
        self.entries.clear();
        self.approx_size_bytes = 0;
    }
}
